presto-main/src/main/java/io/prestosql/operator/scalar/BitwiseFunctions.java

```diff
@@ -75,4 +85,169 @@ public static long bitwiseXor(@SqlType(StandardTypes.BIGINT) long left, @SqlType
+    {
+        if (shift >= 64) {
+            return 0L;
+        }
+        long shifted = (value << shift);
+        return preserveSign(shifted, INTEGER_MASK, INTEGER_SIGNED_BIT);
+    }
+
+    @Description("bitwise left shift")
+    @ScalarFunction("bitwise_left_shift")
+    @SqlType(StandardTypes.BIGINT)
+    public static long bitwiseLeftShiftBigint(@SqlType(StandardTypes.BIGINT) long value, @SqlType(StandardTypes.INTEGER) long shift)
+    {
+        if (shift >= 64) {
+            return 0L;
+        }
+        return value << shift;
+    }
+
+    private static long preserveSign(long shiftedValue, long mask, long signedBit)
+    {
+        if ((shiftedValue & signedBit) != 0) {
+            // Preserve the sign in 2's complement format
+            return shiftedValue | ~mask;
+        }
+
+        return shiftedValue & mask;
+    }
+
+    @Description("bitwise logical right shift")
+    @ScalarFunction("bitwise_right_shift")
+    @SqlType(StandardTypes.TINYINT)
+    public static long bitwiseRightShiftTinyint(@SqlType(StandardTypes.TINYINT) long value, @SqlType(StandardTypes.INTEGER) long shift)
+    {
+        if (shift >= 64) {
+            return 0L;
+        }
+        return (value & TINYINT_MASK) >>> shift;
+    }
+
+    @Description("bitwise logical right shift")
+    @ScalarFunction("bitwise_right_shift")
+    @SqlType(StandardTypes.SMALLINT)
+    public static long bitwiseRightShiftSmallint(@SqlType(StandardTypes.SMALLINT) long value, @SqlType(StandardTypes.INTEGER) long shift)
+    {
+        if (shift >= 64) {
+            return 0L;
+        }
+        return (value & SMALLINT_MASK) >>> shift;
+    }
+
+    @Description("bitwise logical right shift")
+    @ScalarFunction("bitwise_right_shift")
+    @SqlType(StandardTypes.INTEGER)
+    public static long bitwiseRightShiftInteger(@SqlType(StandardTypes.INTEGER) long value, @SqlType(StandardTypes.INTEGER) long shift)
+    {
+        if (shift >= 64) {
+            return 0L;
+        }
+        return (value & INTEGER_MASK) >>> shift;
+    }
+
+    @Description("bitwise logical right shift")
+    @ScalarFunction("bitwise_right_shift")
+    @SqlType(StandardTypes.BIGINT)
+    public static long bitwiseRightShiftBigint(@SqlType(StandardTypes.BIGINT) long value, @SqlType(StandardTypes.INTEGER) long shift)
+    {
+        if (shift >= 64) {
+            return 0L;
+        }
+        return value >>> shift;
+    }
+
+    @Description("bitwise arithmetic right shift")
+    @ScalarFunction("bitwise_right_shift_arithmetic")
+    @SqlType(StandardTypes.TINYINT)
+    public static long bitwiseRightShiftArithmeticTinyint(@SqlType(StandardTypes.TINYINT) long value, @SqlType(StandardTypes.INTEGER) long shift)
+    {
+        if (shift >= 64) {
+            if (value >= 0) {
+                return 0L;
+            }
+            else {
+                return -1L;
+            }
+        }
+        return (value | TINYINT_LEFTPAD_MASK) >> shift;
```

Comment:
Isn't `TINYINT_LEFTPAD_MASK` just `~TINYINT_MASK`? In that case, just do:

```suggestion
        return (value | ~TINYINT_MASK) >> shift;
```

(same for all other implementations)

Reply:
Ah, that's true. I'll fix that. Thanks!